src/runtime/hexagon_remote/pipeline_context.h

```diff
@@ -0,0 +1,92 @@
+#ifndef HALIDE_HEXAGON_REMOTE_PIPELINE_CONTEXT_H
+#define HALIDE_HEXAGON_REMOTE_PIPELINE_CONTEXT_H
+
+#include <HalideRuntime.h>
+
+#include <qurt.h>
+
+// We can't control the stack size on the thread which receives our
+// FastRPC calls. To work around this, we make our own thread, and
+// forward the calls to that thread.
+
+typedef int (*pipeline_argv_t)(void **);
+
+class PipelineContext {
+    void *stack;
+    qurt_thread_t thread;
+    qurt_cond_t wakeup_thread;
+    qurt_cond_t wakeup_caller;
+    qurt_mutex_t work_mutex;
+    qurt_mutex_t wakeup_master_mutex;
+
+    // Shared state
+    pipeline_argv_t function;
+    void **args;
+    int result;
+
+    void thread_main() {
+        while (true) {
+            // Lock and wait for work.
+            qurt_mutex_lock(&work_mutex);
+            qurt_cond_wait(&wakeup_thread, &work_mutex);
```

Comment:
Per in-person discussion, I'm fairly sure there is a sleep/wakeup race here. (If the run method calls qurt_cond_signal before this executor thread gets to this wait call, the code will sleep indefinitely. Condition variables do not keep a count of signal calls. Rather if there is a thread waiting when a signal occurs, it is guaranteed to be woken. Correct code always tests the condition one is waiting for under the exclusion of the mutex before calling cond_wait. This is done in a loop as spurious wakeups are allowed. I usually think of it as if one just wrote a polling loop on some condition with a release/reacquire of the mutex in between. Then the condition variable is added to save CPU cycles. The condition variable logic guarantees atomicity between sleeping and releasing the mutex.)